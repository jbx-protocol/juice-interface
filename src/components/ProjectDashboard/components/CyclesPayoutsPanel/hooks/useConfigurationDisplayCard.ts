import { t } from '@lingui/macro'
import { useMemo } from 'react'

export const useConfigurationDisplayCard = (type: 'current' | 'upcoming') => {
  const title = useMemo(() => {
    if (type === 'current') {
      return t`Current cycle`
    }
    return t`Upcoming cycle`
  }, [type])

  return {
    title,
  }
}
