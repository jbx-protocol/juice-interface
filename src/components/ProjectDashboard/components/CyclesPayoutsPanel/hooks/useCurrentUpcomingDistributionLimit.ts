import {
  useProjectContext,
  useProjectMetadata,
} from 'components/ProjectDashboard/hooks'
import useProjectDistributionLimit from 'hooks/v2v3/contractReader/useProjectDistributionLimit'
import { useProjectUpcomingFundingCycle } from 'hooks/v2v3/contractReader/useProjectUpcomingFundingCycle'

export const useCurrentUpcomingDistributionLimit = (
  type: 'current' | 'upcoming',
) => {
  const { projectId } = useProjectMetadata()
  const {
    primaryETHTerminal,
    distributionLimit: currentDistributionLimit,
    distributionLimitCurrency: currentDistributionLimitCurrency,
  } = useProjectContext()

  const {
    data: upcomingFundingCycleData,
    loading: upcomingFundingCycleLoading,
  } = useProjectUpcomingFundingCycle({ projectId })
  const [upcomingFundingCycle] = upcomingFundingCycleData ?? []

  const {
    data: upcomingDistributionLimitData,
    loading: upcomingDistributionLimitLoading,
  } = useProjectDistributionLimit({
    projectId,
    configuration: upcomingFundingCycle?.configuration.toString(),
    terminal: primaryETHTerminal,
  })

  const [upcomingDistributionLimit, upcomingDistributionLimitCurrency] =
    upcomingDistributionLimitData ?? []

  if (type === 'current') {
    return {
      distributionLimit: currentDistributionLimit,
      distributionLimitCurrency: currentDistributionLimitCurrency,
      loading: false,
    }
  }
  return {
    distributionLimit: upcomingDistributionLimit,
    distributionLimitCurrency: upcomingDistributionLimitCurrency,
    loading: upcomingFundingCycleLoading || upcomingDistributionLimitLoading,
  }
}
