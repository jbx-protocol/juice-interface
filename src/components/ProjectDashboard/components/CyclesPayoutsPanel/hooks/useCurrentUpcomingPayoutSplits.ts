import {
  useProjectContext,
  useProjectMetadata,
} from 'components/ProjectDashboard/hooks'
import { ETH_PAYOUT_SPLIT_GROUP } from 'constants/splits'
import useProjectSplits from 'hooks/v2v3/contractReader/useProjectSplits'
import { useProjectUpcomingFundingCycle } from 'hooks/v2v3/contractReader/useProjectUpcomingFundingCycle'

export const useCurrentUpcomingPayoutSplits = (
  type: 'current' | 'upcoming',
) => {
  const { projectId } = useProjectMetadata()
  const { payoutSplits: currentPayoutSplits } = useProjectContext()

  const {
    data: upcomingFundingCycleData,
    loading: upcomingFundingCycleLoading,
  } = useProjectUpcomingFundingCycle({ projectId })
  const [upcomingFundingCycle] = upcomingFundingCycleData ?? []

  const { data: upcomingPayoutSplits, loading: upcomingProjectSplitsLoading } =
    useProjectSplits({
      projectId,
      splitGroup: ETH_PAYOUT_SPLIT_GROUP,
      domain: upcomingFundingCycle?.configuration?.toString(),
    })

  if (type === 'current') {
    return { splits: currentPayoutSplits, loading: false }
  }
  return {
    splits: upcomingPayoutSplits,
    loading: upcomingFundingCycleLoading || upcomingProjectSplitsLoading,
  }
}
