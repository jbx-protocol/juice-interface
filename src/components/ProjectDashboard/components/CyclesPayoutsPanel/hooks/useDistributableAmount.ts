import { V2V3ProjectContext } from 'contexts/v2v3/Project/V2V3ProjectContext'
import { BigNumber } from 'ethers'
import { V2V3CurrencyOption } from 'models/v2v3/currencyOption'
import { useContext } from 'react'
import { V2V3_CURRENCY_ETH } from 'utils/v2v3/currency'

export const useDistributableAmount = () => {
  const {
    distributionLimit,
    usedDistributionLimit,
    balanceInDistributionLimitCurrency,
    distributionLimitCurrency,
  } = useContext(V2V3ProjectContext)

  const currency =
    distributionLimitCurrency?.toNumber() ??
    (V2V3_CURRENCY_ETH as V2V3CurrencyOption)

  const effectiveDistributionLimit = distributionLimit ?? BigNumber.from(0)
  const distributedAmount = usedDistributionLimit ?? BigNumber.from(0)
  const balInDistributionLimit =
    balanceInDistributionLimitCurrency ?? BigNumber.from(0)

  const distributable = effectiveDistributionLimit.eq(0)
    ? effectiveDistributionLimit
    : effectiveDistributionLimit.sub(distributedAmount)

  const distributableAmount = balInDistributionLimit.gt(distributable)
    ? distributable
    : balInDistributionLimit

  return {
    distributableAmount,
    currency,
  }
}
