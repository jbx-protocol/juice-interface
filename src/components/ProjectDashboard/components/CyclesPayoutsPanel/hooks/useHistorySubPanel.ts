import {
  useProjectContext,
  useProjectMetadata,
} from 'components/ProjectDashboard/hooks'
import { fetchPastFundingCycles } from 'components/v2v3/V2V3Project/V2V3FundingCycleSection/FundingCycleHistory/utils'
import { ETH_TOKEN_ADDRESS } from 'constants/v2v3/juiceboxTokens'
import { V2V3ProjectContractsContext } from 'contexts/v2v3/ProjectContracts/V2V3ProjectContractsContext'
import { BigNumber, Contract } from 'ethers'
import { JB_CONTROLLER_V_3_1 } from 'hooks/v2v3/V2V3ProjectContracts/projectContractLoaders/useProjectController'
import { V2V3CurrencyOption } from 'models/v2v3/currencyOption'
import { V2V3FundingCycle } from 'models/v2v3/fundingCycle'
import moment from 'moment'
import { useCallback, useContext, useEffect, useReducer } from 'react'
import { isBigNumberish } from 'utils/bigNumbers'
import { callContract } from 'utils/callContract'
import { fromWad } from 'utils/format/formatNumber'
import { formatCurrencyAmount } from 'utils/formatCurrencyAmount'
import { HistoryData } from '../components/HistorySubPanel'

type State = {
  loading: boolean
  data: HistoryData
  error?: string
}

type Action =
  | { type: 'request' }
  | { type: 'success'; payload: HistoryData }
  | { type: 'failure'; payload: string }

const reducer = (state: State, action: Action) => {
  switch (action.type) {
    case 'request':
      return { ...state, loading: true, error: undefined }
    case 'success':
      return { ...state, loading: false, data: action.payload }
    case 'failure':
      return { ...state, loading: false, error: action.payload }
  }
}

export const useHistorySubPanel = () => {
  const { projectId } = useProjectMetadata()
  const { fundingCycle, primaryETHTerminal } = useProjectContext()
  const {
    contracts,
    versions: { JBControllerVersion },
  } = useContext(V2V3ProjectContractsContext)
  const [state, dispatch] = useReducer(reducer, {
    loading: false,
    data: [],
  })

  const fetchCurrencyOfFundingCycle = useCallback(
    async (fundingCycle: V2V3FundingCycle) => {
      const terminal = primaryETHTerminal
      const { JBController, JBFundAccessConstraintsStore } = contracts
      const contract =
        JBControllerVersion === JB_CONTROLLER_V_3_1
          ? JBFundAccessConstraintsStore
          : JBController
      const result = await callContract({
        contract,
        contracts: contracts as Record<string, Contract>,
        functionName: 'distributionLimitOf',
        args:
          projectId && fundingCycle.configuration && terminal && contract
            ? [
                projectId,
                fundingCycle.configuration.toString(),
                terminal,
                ETH_TOKEN_ADDRESS,
              ]
            : null,
      })
      if (typeof result === undefined) return
      if (!(Array.isArray(result) && result.length === 2)) {
        console.error('Unexpected result from distributionLimitOf', result)
        throw new Error('Unexpected result from distributionLimitOf')
      }
      const [, currency] = result
      if (!isBigNumberish(currency)) {
        console.error('Unexpected result from distributionLimitOf', result)
        throw new Error('Unexpected result from distributionLimitOf')
      }
      const currencyOption = BigNumber.from(currency).toNumber()
      if (currencyOption === 0) {
        return undefined
      }
      return BigNumber.from(currency).toNumber() as V2V3CurrencyOption
    },
    [JBControllerVersion, contracts, primaryETHTerminal, projectId],
  )

  const fetchUsedDistributionLimitOfFundingCycle = useCallback(
    async (fundingCycle: V2V3FundingCycle) => {
      const terminal = primaryETHTerminal
      const { number } = fundingCycle
      const result = await callContract({
        contract: contracts.JBETHPaymentTerminalStore,
        contracts: contracts as Record<string, Contract>,
        functionName: 'usedDistributionLimitOf',
        args:
          terminal && projectId && number
            ? [terminal, projectId, number]
            : null,
      })
      if (typeof result === undefined) return
      if (!isBigNumberish(result)) {
        console.error('Unexpected result from usedDistributionLimitOf')
        throw new Error('Unexpected result from usedDistributionLimitOf')
      }
      return Number(fromWad(BigNumber.from(result)))
    },
    [contracts, primaryETHTerminal, projectId],
  )

  const fetchWithdrawnOfFundingCycle = useCallback(
    async (fundingCycle: V2V3FundingCycle) => {
      const [withdrawn, currency] = await Promise.all([
        fetchUsedDistributionLimitOfFundingCycle(fundingCycle),
        fetchCurrencyOfFundingCycle(fundingCycle),
      ])
      return formatCurrencyAmount({ amount: withdrawn, currency }) ?? '0Îž'
    },
    [fetchCurrencyOfFundingCycle, fetchUsedDistributionLimitOfFundingCycle],
  )

  useEffect(() => {
    dispatch({ type: 'request' })

    const fetchHistory = async () => {
      if (!projectId || !fundingCycle || !contracts.JBController) return
      try {
        const data = await fetchPastFundingCycles({
          projectId,
          currentFundingCycle: fundingCycle,
          JBController: contracts.JBController,
        })

        const result = await Promise.all(
          data.map(async d => ({
            cycleNumber: d[0].number.toString(),
            withdrawn: await fetchWithdrawnOfFundingCycle(d[0]),
            date: `${moment(
              d[0].start.add(d[0].duration).mul(1000).toNumber(),
            ).fromNow(true)} ago`,
            _metadata: {
              fundingCycle: d[0],
              metadata: d[1],
            },
          })),
        )

        dispatch({ type: 'success', payload: result })
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
      } catch (error: any) {
        console.error('error', error)
        dispatch({ type: 'failure', payload: error.message })
      }
    }

    fetchHistory()
  }, [
    contracts.JBController,
    fetchCurrencyOfFundingCycle,
    fetchWithdrawnOfFundingCycle,
    fundingCycle,
    projectId,
  ])

  return state
}
