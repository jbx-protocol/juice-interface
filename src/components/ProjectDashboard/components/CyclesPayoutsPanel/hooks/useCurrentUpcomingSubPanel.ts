import { t } from '@lingui/macro'
import {
  useProjectContext,
  useProjectMetadata,
} from 'components/ProjectDashboard/hooks'
import { useFundingCycleCountdown } from 'components/ProjectDashboard/hooks/useFundingCycleCountdown'
import { timeSecondsToDateString } from 'components/ProjectDashboard/utils/timeSecondsToDateString'
import { useProjectUpcomingFundingCycle } from 'hooks/v2v3/contractReader/useProjectUpcomingFundingCycle'
import { useMemo } from 'react'

export const useCurrentUpcomingSubPanel = (type: 'current' | 'upcoming') => {
  const { projectId } = useProjectMetadata()
  const {
    fundingCycle,
    loading: { fundingCycleLoading },
  } = useProjectContext()
  const {
    data: upcomingFundingCycleData,
    loading: upcomingFundingCycleLoading,
  } = useProjectUpcomingFundingCycle({ projectId })
  const [upcomingFundingCycle] = upcomingFundingCycleData ?? []
  const { timeRemainingText } = useFundingCycleCountdown()

  const cycleNumber = useMemo(() => {
    if (type === 'current') {
      return fundingCycle?.number.toNumber()
    }
    return fundingCycle?.number ? fundingCycle.number.toNumber() + 1 : undefined
  }, [fundingCycle?.number, type])

  const cycleUnlocked = useMemo(() => {
    if (type === 'current') {
      return fundingCycle?.duration?.isZero() ?? true
    }
    return upcomingFundingCycle?.duration?.isZero() ?? true
  }, [fundingCycle?.duration, type, upcomingFundingCycle?.duration])

  const upcomingCycleLength = useMemo(() => {
    if (!upcomingFundingCycle) return
    if (cycleUnlocked) return '-'
    return timeSecondsToDateString(
      upcomingFundingCycle.duration.toNumber(),
      'short',
    )
  }, [cycleUnlocked, upcomingFundingCycle])

  const status = useMemo(() => {
    if (cycleUnlocked) return t`Unlocked`
    return t`Locked`
  }, [cycleUnlocked])

  const remainingTime = useMemo(() => {
    if (cycleUnlocked) return '-'
    return timeRemainingText
  }, [cycleUnlocked, timeRemainingText])

  // Short circuit current for faster loading
  if (type === 'current') {
    if (fundingCycleLoading) return { loading: true, type }
    return {
      loading: false,
      type,
      cycleNumber,
      status,
      remainingTime,
    }
  }

  if (fundingCycleLoading || upcomingFundingCycleLoading)
    return {
      loading: true,
      type,
    }

  return {
    loading: false,
    type,
    cycleNumber,
    status,
    cycleLength: upcomingCycleLength,
  }
}
