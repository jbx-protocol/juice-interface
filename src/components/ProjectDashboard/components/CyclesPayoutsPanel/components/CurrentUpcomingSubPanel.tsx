import { Trans, t } from '@lingui/macro'
import { useMemo } from 'react'
import { twMerge } from 'tailwind-merge'
import { TitleDescriptionDisplayCard } from '../../ui/TitleDescriptionDisplayCard'
import { useCurrentUpcomingSubPanel } from '../hooks/useCurrentUpcomingSubPanel'
import { ConfigurationDisplayCard } from './ConfigurationDisplayCard'
import { currentCycleRemainingLengthTooltip } from './CyclesPanelTooltips'
import { PayoutsSubPanel } from './PayoutsSubPanel'
import { UpcomingCycleChangesCallout } from './UpcomingCycleChangesCallout'

const CYCLE_NUMBER_INDEX = 0
const STATUS_INDEX = 1
const CYCLE_LENGTH_INDEX = 2

export const CurrentUpcomingSubPanel = ({
  id,
}: {
  id: 'current' | 'upcoming'
}) => {
  const info = useCurrentUpcomingSubPanel(id)

  const topPanelsInfo = useMemo(() => {
    const topPanelInfo = [
      {
        title: t`Cycle #`,
        value: info.cycleNumber,
      },
      {
        title: t`Status`,
        value: info.status,
      },
    ]
    if (info.type === 'current') {
      topPanelInfo.push({
        title: t`Remaining time`,
        value: info.remainingTime,
      })
    }
    if (info.type === 'upcoming') {
      topPanelInfo.push({
        title: t`Cycle length`,
        value: info.cycleLength,
      })
    }
    return topPanelInfo
  }, [
    info.cycleLength,
    info.cycleNumber,
    info.remainingTime,
    info.status,
    info.type,
  ])

  const cycleLengthTooltip =
    info.type === 'current' ? currentCycleRemainingLengthTooltip : undefined

  const cycleStatusValue = topPanelsInfo[STATUS_INDEX].value

  const cycleStatusTooltip = (
    <Trans>
      This project's rules will be locked in place for {cycleStatusValue} days.
    </Trans>
  )

  return (
    <div>
      <div className="flex flex-col gap-4">
        {id === 'upcoming' && <UpcomingCycleChangesCallout />}

        <div className="grid grid-cols-2 gap-4 md:flex">
          <TitleDescriptionDisplayCard
            className="w-full md:max-w-[127px]"
            title={topPanelsInfo[CYCLE_NUMBER_INDEX].title}
            description={
              topPanelsInfo[CYCLE_NUMBER_INDEX].value ?? <Skeleton />
            }
          />
          <TitleDescriptionDisplayCard
            className="w-full md:w-fit"
            title={topPanelsInfo[STATUS_INDEX].title}
            description={
              topPanelsInfo[STATUS_INDEX].value ?? <Skeleton className="w-22" />
            }
            tooltip={cycleStatusTooltip}
          />
          <TitleDescriptionDisplayCard
            className="col-span-2 md:flex-1"
            title={topPanelsInfo[CYCLE_LENGTH_INDEX].title}
            description={
              topPanelsInfo[CYCLE_LENGTH_INDEX].value ?? (
                <Skeleton className="w-40" />
              )
            }
            tooltip={cycleLengthTooltip}
          />
        </div>
        <ConfigurationDisplayCard type={info.type} />
      </div>

      <PayoutsSubPanel className="mt-12" type={id} />
    </div>
  )
}

const Skeleton = ({ className }: { className?: string }) => (
  <div
    className={twMerge(
      'h-8 w-12 animate-pulse rounded-lg bg-smoke-200 dark:bg-slate-500',
      className,
    )}
  />
)
