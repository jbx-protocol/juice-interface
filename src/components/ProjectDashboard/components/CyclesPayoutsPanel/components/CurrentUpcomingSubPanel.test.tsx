/**
 * @jest-environment jsdom
 */
import { render } from '@testing-library/react'
import { useCurrentUpcomingSubPanel } from '../hooks/useCurrentUpcomingSubPanel'
import { CurrentUpcomingSubPanel } from './CurrentUpcomingSubPanel'

jest.mock('./ConfigurationDisplayCard', () => ({
  ConfigurationDisplayCard: jest.fn(() => <div>Configuration</div>),
}))

jest.mock('../hooks/useCurrentUpcomingSubPanel', () => ({
  useCurrentUpcomingSubPanel: jest.fn(),
}))

jest.mock('./PayoutsSubPanel', () => ({
  PayoutsSubPanel: jest.fn(() => <div>Payouts</div>),
}))

const DefaultResponse = {
  loading: false,
  cycleNumber: 1,
  status: 'Active',
  remainingTime: '1 day',
  cycleLength: '1 day',
  type: 'current',
}

const mockUseCurrentUpcomingSubPanel = useCurrentUpcomingSubPanel as jest.Mock

describe('CurrentUpcomingSubPanel', () => {
  beforeEach(() => {
    jest.clearAllMocks()
    mockUseCurrentUpcomingSubPanel.mockImplementation(() => DefaultResponse)
  })

  it.each(['current', 'upcoming'] as const)('renders %p', id => {
    mockUseCurrentUpcomingSubPanel.mockImplementationOnce(() => ({
      ...DefaultResponse,
      type: id,
      loading: false,
    }))
    const { container } = render(<CurrentUpcomingSubPanel id={id} />)
    expect(container).toMatchSnapshot()
  })

  it.each(['current', 'upcoming'] as const)(
    'renders a skeleton when loading %p',
    id => {
      mockUseCurrentUpcomingSubPanel.mockReturnValueOnce({
        ...DefaultResponse,
        type: id,
        loading: true,
      })
      const { container } = render(<CurrentUpcomingSubPanel id={id} />)
      expect(container).toMatchSnapshot()
    },
  )

  it('renders remaining time when current', () => {
    const { getByText } = render(<CurrentUpcomingSubPanel id="current" />)
    expect(getByText('Remaining time')).toBeInTheDocument()
  })

  it('renders cycle length when upcoming', () => {
    mockUseCurrentUpcomingSubPanel.mockReturnValueOnce({
      ...DefaultResponse,
      type: 'upcoming',
      loading: false,
    })
    const { getByText } = render(<CurrentUpcomingSubPanel id="upcoming" />)
    expect(getByText('Cycle length')).toBeInTheDocument()
  })
})
