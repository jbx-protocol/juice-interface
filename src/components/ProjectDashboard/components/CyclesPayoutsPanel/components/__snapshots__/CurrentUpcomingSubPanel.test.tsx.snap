// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrentUpcomingSubPanel renders "current" 1`] = `
<div>
  <div>
    <div
      class="flex flex-col gap-4"
    >
      <div
        class="grid grid-cols-2 gap-4 md:flex"
      >
        <div
          class="rounded-lg bg-smoke-50 py-5 px-6 dark:bg-slate-700 w-full md:max-w-[127px]"
        >
          <div>
            <div
              class="flex flex-1 items-center justify-between"
            >
              <span
                class="inline-flex max-w-min items-center gap-1 whitespace-nowrap font-body text-sm font-medium text-grey-600 dark:text-slate-200"
              >
                <span>
                  Cycle #
                </span>
              </span>
            </div>
            <div
              class="mt-2 truncate font-heading text-xl font-medium dark:text-slate-50"
            >
              1
            </div>
          </div>
        </div>
        <div
          class="rounded-lg bg-smoke-50 py-5 px-6 dark:bg-slate-700 w-full md:w-fit"
        >
          <div>
            <div
              class="flex flex-1 items-center justify-between"
            >
              <span
                class="inline-flex max-w-min items-center gap-1 whitespace-nowrap font-body text-sm font-medium text-grey-600 dark:text-slate-200"
              >
                <span
                  class="leading-none"
                >
                  Status
                </span>
                <svg
                  aria-hidden="true"
                  class="h-4 w-4 text-grey-500 dark:text-slate-200"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M9.879 7.519c1.171-1.025 3.071-1.025 4.242 0 1.172 1.025 1.172 2.687 0 3.712-.203.179-.43.326-.67.442-.745.361-1.45.999-1.45 1.827v.75M21 12a9 9 0 11-18 0 9 9 0 0118 0zm-9 5.25h.008v.008H12v-.008z"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
              </span>
            </div>
            <div
              class="mt-2 truncate font-heading text-xl font-medium dark:text-slate-50"
            >
              Active
            </div>
          </div>
        </div>
        <div
          class="rounded-lg bg-smoke-50 py-5 px-6 dark:bg-slate-700 col-span-2 md:flex-1"
        >
          <div>
            <div
              class="flex flex-1 items-center justify-between"
            >
              <span
                class="inline-flex max-w-min items-center gap-1 whitespace-nowrap font-body text-sm font-medium text-grey-600 dark:text-slate-200"
              >
                <span
                  class="leading-none"
                >
                  Remaining time
                </span>
                <svg
                  aria-hidden="true"
                  class="h-4 w-4 text-grey-500 dark:text-slate-200"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M9.879 7.519c1.171-1.025 3.071-1.025 4.242 0 1.172 1.025 1.172 2.687 0 3.712-.203.179-.43.326-.67.442-.745.361-1.45.999-1.45 1.827v.75M21 12a9 9 0 11-18 0 9 9 0 0118 0zm-9 5.25h.008v.008H12v-.008z"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
              </span>
            </div>
            <div
              class="mt-2 truncate font-heading text-xl font-medium dark:text-slate-50"
            >
              1 day
            </div>
          </div>
        </div>
      </div>
      <div>
        Configuration
      </div>
    </div>
    <div>
      Payouts
    </div>
  </div>
</div>
`;

exports[`CurrentUpcomingSubPanel renders "upcoming" 1`] = `
<div>
  <div>
    <div
      class="flex flex-col gap-4"
    >
      <div
        class="grid grid-cols-2 gap-4 md:flex"
      >
        <div
          class="rounded-lg bg-smoke-50 py-5 px-6 dark:bg-slate-700 w-full md:max-w-[127px]"
        >
          <div>
            <div
              class="flex flex-1 items-center justify-between"
            >
              <span
                class="inline-flex max-w-min items-center gap-1 whitespace-nowrap font-body text-sm font-medium text-grey-600 dark:text-slate-200"
              >
                <span>
                  Cycle #
                </span>
              </span>
            </div>
            <div
              class="mt-2 truncate font-heading text-xl font-medium dark:text-slate-50"
            >
              1
            </div>
          </div>
        </div>
        <div
          class="rounded-lg bg-smoke-50 py-5 px-6 dark:bg-slate-700 w-full md:w-fit"
        >
          <div>
            <div
              class="flex flex-1 items-center justify-between"
            >
              <span
                class="inline-flex max-w-min items-center gap-1 whitespace-nowrap font-body text-sm font-medium text-grey-600 dark:text-slate-200"
              >
                <span
                  class="leading-none"
                >
                  Status
                </span>
                <svg
                  aria-hidden="true"
                  class="h-4 w-4 text-grey-500 dark:text-slate-200"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M9.879 7.519c1.171-1.025 3.071-1.025 4.242 0 1.172 1.025 1.172 2.687 0 3.712-.203.179-.43.326-.67.442-.745.361-1.45.999-1.45 1.827v.75M21 12a9 9 0 11-18 0 9 9 0 0118 0zm-9 5.25h.008v.008H12v-.008z"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
              </span>
            </div>
            <div
              class="mt-2 truncate font-heading text-xl font-medium dark:text-slate-50"
            >
              Active
            </div>
          </div>
        </div>
        <div
          class="rounded-lg bg-smoke-50 py-5 px-6 dark:bg-slate-700 col-span-2 md:flex-1"
        >
          <div>
            <div
              class="flex flex-1 items-center justify-between"
            >
              <span
                class="inline-flex max-w-min items-center gap-1 whitespace-nowrap font-body text-sm font-medium text-grey-600 dark:text-slate-200"
              >
                <span>
                  Cycle length
                </span>
              </span>
            </div>
            <div
              class="mt-2 truncate font-heading text-xl font-medium dark:text-slate-50"
            >
              1 day
            </div>
          </div>
        </div>
      </div>
      <div>
        Configuration
      </div>
    </div>
    <div>
      Payouts
    </div>
  </div>
</div>
`;

exports[`CurrentUpcomingSubPanel renders a skeleton when loading "current" 1`] = `
<div>
  <div>
    <div
      class="flex flex-col gap-4"
    >
      <div
        class="grid grid-cols-2 gap-4 md:flex"
      >
        <div
          class="rounded-lg bg-smoke-50 py-5 px-6 dark:bg-slate-700 w-full md:max-w-[127px]"
        >
          <div>
            <div
              class="flex flex-1 items-center justify-between"
            >
              <span
                class="inline-flex max-w-min items-center gap-1 whitespace-nowrap font-body text-sm font-medium text-grey-600 dark:text-slate-200"
              >
                <span>
                  Cycle #
                </span>
              </span>
            </div>
            <div
              class="mt-2 truncate font-heading text-xl font-medium dark:text-slate-50"
            >
              1
            </div>
          </div>
        </div>
        <div
          class="rounded-lg bg-smoke-50 py-5 px-6 dark:bg-slate-700 w-full md:w-fit"
        >
          <div>
            <div
              class="flex flex-1 items-center justify-between"
            >
              <span
                class="inline-flex max-w-min items-center gap-1 whitespace-nowrap font-body text-sm font-medium text-grey-600 dark:text-slate-200"
              >
                <span
                  class="leading-none"
                >
                  Status
                </span>
                <svg
                  aria-hidden="true"
                  class="h-4 w-4 text-grey-500 dark:text-slate-200"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M9.879 7.519c1.171-1.025 3.071-1.025 4.242 0 1.172 1.025 1.172 2.687 0 3.712-.203.179-.43.326-.67.442-.745.361-1.45.999-1.45 1.827v.75M21 12a9 9 0 11-18 0 9 9 0 0118 0zm-9 5.25h.008v.008H12v-.008z"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
              </span>
            </div>
            <div
              class="mt-2 truncate font-heading text-xl font-medium dark:text-slate-50"
            >
              Active
            </div>
          </div>
        </div>
        <div
          class="rounded-lg bg-smoke-50 py-5 px-6 dark:bg-slate-700 col-span-2 md:flex-1"
        >
          <div>
            <div
              class="flex flex-1 items-center justify-between"
            >
              <span
                class="inline-flex max-w-min items-center gap-1 whitespace-nowrap font-body text-sm font-medium text-grey-600 dark:text-slate-200"
              >
                <span
                  class="leading-none"
                >
                  Remaining time
                </span>
                <svg
                  aria-hidden="true"
                  class="h-4 w-4 text-grey-500 dark:text-slate-200"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M9.879 7.519c1.171-1.025 3.071-1.025 4.242 0 1.172 1.025 1.172 2.687 0 3.712-.203.179-.43.326-.67.442-.745.361-1.45.999-1.45 1.827v.75M21 12a9 9 0 11-18 0 9 9 0 0118 0zm-9 5.25h.008v.008H12v-.008z"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
              </span>
            </div>
            <div
              class="mt-2 truncate font-heading text-xl font-medium dark:text-slate-50"
            >
              1 day
            </div>
          </div>
        </div>
      </div>
      <div>
        Configuration
      </div>
    </div>
    <div>
      Payouts
    </div>
  </div>
</div>
`;

exports[`CurrentUpcomingSubPanel renders a skeleton when loading "upcoming" 1`] = `
<div>
  <div>
    <div
      class="flex flex-col gap-4"
    >
      <div
        class="grid grid-cols-2 gap-4 md:flex"
      >
        <div
          class="rounded-lg bg-smoke-50 py-5 px-6 dark:bg-slate-700 w-full md:max-w-[127px]"
        >
          <div>
            <div
              class="flex flex-1 items-center justify-between"
            >
              <span
                class="inline-flex max-w-min items-center gap-1 whitespace-nowrap font-body text-sm font-medium text-grey-600 dark:text-slate-200"
              >
                <span>
                  Cycle #
                </span>
              </span>
            </div>
            <div
              class="mt-2 truncate font-heading text-xl font-medium dark:text-slate-50"
            >
              1
            </div>
          </div>
        </div>
        <div
          class="rounded-lg bg-smoke-50 py-5 px-6 dark:bg-slate-700 w-full md:w-fit"
        >
          <div>
            <div
              class="flex flex-1 items-center justify-between"
            >
              <span
                class="inline-flex max-w-min items-center gap-1 whitespace-nowrap font-body text-sm font-medium text-grey-600 dark:text-slate-200"
              >
                <span
                  class="leading-none"
                >
                  Status
                </span>
                <svg
                  aria-hidden="true"
                  class="h-4 w-4 text-grey-500 dark:text-slate-200"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M9.879 7.519c1.171-1.025 3.071-1.025 4.242 0 1.172 1.025 1.172 2.687 0 3.712-.203.179-.43.326-.67.442-.745.361-1.45.999-1.45 1.827v.75M21 12a9 9 0 11-18 0 9 9 0 0118 0zm-9 5.25h.008v.008H12v-.008z"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
              </span>
            </div>
            <div
              class="mt-2 truncate font-heading text-xl font-medium dark:text-slate-50"
            >
              Active
            </div>
          </div>
        </div>
        <div
          class="rounded-lg bg-smoke-50 py-5 px-6 dark:bg-slate-700 col-span-2 md:flex-1"
        >
          <div>
            <div
              class="flex flex-1 items-center justify-between"
            >
              <span
                class="inline-flex max-w-min items-center gap-1 whitespace-nowrap font-body text-sm font-medium text-grey-600 dark:text-slate-200"
              >
                <span>
                  Cycle length
                </span>
              </span>
            </div>
            <div
              class="mt-2 truncate font-heading text-xl font-medium dark:text-slate-50"
            >
              1 day
            </div>
          </div>
        </div>
      </div>
      <div>
        Configuration
      </div>
    </div>
    <div>
      Payouts
    </div>
  </div>
</div>
`;
